File: 백준/Gold/9576. 책 나눠주기/책 나눠주기.py
import sys
input = sys.stdin.readline

T = int(input())
for _ in range(T):
    N, M = map(int, input().split())
    requests = []
    
    for _ in range(M):
        request = list(map(int, input().split()))
        requests.append((request[0], request[1]))
        
    sorted_requests = sorted(requests, key=lambda x: (x[1], x[0])) 
    
    used = [False] * (N + 1)
    max_students = 0
    
    for start, end in sorted_requests:
        for book in range(start, end + 1):
            if not used[book]:
                max_students += 1
                used[book] = True
                break
                
    print(max_students)

File: 백준/Gold/9576. 책 나눠주기/책 나눠주기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int T = Integer.parseInt(st.nextToken());
        
        for (int t = 0; t < T; t++) {
            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());
            
            int[][] requests = new int[M][2];
            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                requests[i][0] = Integer.parseInt(st.nextToken());
                requests[i][1] = Integer.parseInt(st.nextToken());
            }
            Arrays.sort(requests, (a, b) -> Integer.compare(a[1], b[1]));
            
            boolean[] used = new boolean[N + 1];
            int maxStudents = 0;
            
            for (int[] request: requests) {
                int start = request[0];
                int end = request[1];
                for (int book = start; book <= end; book++) {
                    if (!used[book]) {
                        maxStudents++;
                        used[book] = true;
                        break;
                    }
                }
            }
            System.out.println(maxStudents);
        }
    }
}